client/src/main/java/io/pravega/client/stream/impl/ControllerImpl.java

```diff
@@ -601,21 +602,42 @@ private Boolean handleScaleResponse(Stream stream, ScaleResponse response) {
     }
     
     @Override
-    public CompletableFuture<Set<Segment>> getSuccessors(StreamCut from) {
+    public CompletableFuture<Set<Segment>> getSuccessors(final StreamCut from) {
         Exceptions.checkNotClosed(closed.get(), this);
         Stream stream = from.asImpl().getStream();
         long traceId = LoggerHelpers.traceEnter(log, "getSuccessorsFromCut", stream);
-        HashSet<Segment> unread = new HashSet<>(from.asImpl().getPositions().keySet());
         val currentSegments = getAndHandleExceptions(getCurrentSegments(stream.getScope(), stream.getStreamName()),
-                                                     RuntimeException::new);
-        unread.addAll(computeKnownUnreadSegments(currentSegments, from));
+                                                     RuntimeException::new).getSegments();
+        final Set<Segment> unread = getSegmentsInclusive(from.asImpl(), currentSegments);
+        LoggerHelpers.traceLeave(log, "getSuccessorsFromCut", traceId);
+        return CompletableFuture.completedFuture(unread);
+    }
+
+    @Override
+    public CompletableFuture<Set<Segment>> getSegments(final StreamCut fromStreamCut, final StreamCut toStreamCut) {
```

Comment:
Nowhere does this method actually validate that fromStreamCut is strictly below toStreamCut.

Reply:
done, added validation and corresponding unit tests.